api/src/main/java/org/pmiops/workbench/db/dao/DataSetServiceImpl.java

```diff
@@ -132,185 +163,241 @@ public DataSet saveDataSet(
+          .collect(Collectors.toList());
+
+      final ValuesLinkingPair valuesLinkingPair = getValueSelectsAndJoins(
+          valuePairsForCurrentDomain);
+
+      queryBuilder.append(String.join(", ", valuesLinkingPair.getSelects()))
+          .append(" ")
+          .append(formatValuesLinkingPair(valuesLinkingPair));
+      validateSelectedConceptSetsForDomain(conceptSetsSelected);
+
+      // TODO: this wont' work if there aren't any
+      final String conceptSetIDs = conceptSetsSelected.stream()
+          .filter(cs -> domain == cs.getDomainEnum())
+          .flatMap(cs -> cs.getConceptIds().stream().map(cid -> Long.toString(cid)))
+          .collect(Collectors.joining(", "));
+      final String conceptSetListQuery = " IN (" + conceptSetIDs + ")";
+
+      final Optional<DomainConceptIdInfo> domainConceptIdsMaybe =
+          bigQuerySchemaConfig.cohortTables.values().stream()
+              .filter(config -> domain.toString().equals(config.domain))
+              .map(
+                  tableConfig ->
+                      new DomainConceptIdInfo(
+                          getColumnName(tableConfig, "source"),
+                          getColumnName(tableConfig, "standard")))
+              .findFirst();
+
+      final DomainConceptIdInfo domainConceptIdInfo = domainConceptIdsMaybe.orElseThrow(
+          () -> new ServerErrorException(String.format(
+              "Couldn't find source and standard columns for domain: %s",
+              domain.toString())));
+
+      // This adds the where clauses for cohorts and concept sets.
+      queryBuilder.append(
+          formatWhereClause(conceptSetListQuery, " WHERE \n(",
+              domainConceptIdInfo.getStandardConceptIdColumn(), " OR \n",
+              domainConceptIdInfo.getSourceConceptIdColumn(),
+              conceptSetListQuery, ")"));
+      if (!includesAllParticipants) {
+        queryBuilder
+            .append(" \nAND (PERSON_ID IN (")
+            .append(cohortQueries)
+            .append("))");
       }
 
-      queryMap.put(query, cohortParameters);
-      QueryJobConfiguration queryJobConfiguration =
+      final String completeQuery = queryBuilder.toString();
+
+      final QueryJobConfiguration queryJobConfiguration =
           bigQueryService.filterBigQueryConfig(
-              QueryJobConfiguration.newBuilder(query)
+              QueryJobConfiguration.newBuilder(completeQuery)
                   .setNamedParameters(cohortParameters)
                   .setUseLegacySql(false)
                   .build());
-      dataSetUtil.put(d.toString(), queryJobConfiguration);
+      resultBuilder.put(domain.toString(), queryJobConfiguration);
     }
-    return dataSetUtil;
+    return resultBuilder.build();
+  }
+
+  private static String formatWhereClause(String conceptSetListQuery, String s,
+      String standardConceptIdColumn, String s2, String sourceConceptIdColumn,
+      String conceptSetListQuery2, String s3) {
+    return s
```

Comment:
Should this be string builder rather than concatenation?

Reply:
Yes. I wound up re-inlining this function. IntelliJ was too greedy in parameterizing it. This should be going into my phase 2 class, along with anything that tries to write sql.